### PR TITLE
Add deprecated function warning

### PR DESCRIPTION
**Pull request by oneonestar**
Generate a warning when a query is using a `@Deprecated` function.
This works for annotated `Scalar`, `Parametric Scalar`, `Aggregation`, `Window` functions.
Both built-in UDFs and UDFs from plugin are supported.
Doesn't work for classes that are directly implements `SqlFunction` without using annotation.

This feature can be enabled by the feature flag `analyzer.query-diagnosis-warning-enabled` or `enable_query_diagnosis_warning` session. Default is not enabled.

I have two questions:
* Is the feature flag needed?
* Should this function be default ON / OFF?

```java
@Deprecated
@ScalarFunction("deprecated_function")
public static Slice funcABC() { ... }
```
```sql
presto> SELECT deprecated_function();
.....
(1 row)

WARNING: Detected use of deprecated function: deprecated_function
```

Supersedes #1006

### CODE REVIEW

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java [hunk 1]
```diff
@@ -0,0 +1,107 @@
...
+            new DefaultTraversalVisitor<Void, Void>()
+            {
+                @Override
+                protected Void visitFunctionCall(FunctionCall node, Void context)
+                {
+                    Optional<ResolvedFunction> resolvedFunction = ResolvedFunction.fromQualifiedName(node.getName());
+                    if (resolvedFunction.isPresent()) {
+                        FunctionMetadata functionMetadata = metadata.getFunctionMetadata(resolvedFunction.get());
+                        if (functionMetadata.isDeprecated()) {
+                            warningCollector.add(new PrestoWarning(DREPRCATED_FUNCTION,
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java]: This should be done in the analyzer, not during query optimization. The check should go in `ExpressionAnalyzer.visitFunctionCall`.

File: presto-main/src/main/java/io/prestosql/SystemSessionProperties.java [hunk 2]
```diff
@@ -552,6 +553,11 @@ public SystemSessionProperties(
                         IGNORE_DOWNSTREAM_PREFERENCES,
                         "Ignore Parent's PreferredProperties in AddExchange optimizer",
                         featuresConfig.isIgnoreDownstreamPreferences(),
+                        false),
+                booleanProperty(
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/SystemSessionProperties.java]: What's the purpose of this session property?

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+                                    String.format("Detected use of deprecated function: %s", functionMetadata.getSignature().getName())));
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java]: Ideally, the deprecation message should be able to contain instructions about the reason for deprecation and possible substitutions. So, we'll need to capture this information in the deprecation annotation. Since `@Deprecated` doesn't take a description, we might need an additional annotation.

File: presto-spi/src/main/java/io/prestosql/spi/connector/StandardWarningCode.java [hunk 4]
```diff
@@ -21,6 +21,7 @@
 {
     TOO_MANY_STAGES(0x0000_0001),
     REDUNDANT_ORDER_BY(0x0000_0002),
+    DREPRCATED_FUNCTION(0x0000_0003),
```
Comment (martint) [inline presto-spi/src/main/java/io/prestosql/spi/connector/StandardWarningCode.java]: Typo: DREPRCATED

File: presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java (hunk 3 above)
Comment (oneonestar) [reply presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java]: This requires adding a new annotation and pass it into FunctionMetadata, which looks like a bit overkill to me. The existing `@Description` can be used to explain the substitution, just like what we do in session description:
https://github.com/prestosql/presto/blob/5d013879f430fcb90cf37752c501df16baa073fa/presto-main/src/main/java/io/prestosql/SystemSessionProperties.java#L301

Comment (martint) [reply presto-main/src/main/java/io/prestosql/sql/planner/sanity/warnings/DeprecatedFunctionWarning.java]: That's a good point. Let's use that.

File: presto-docs/src/main/sphinx/develop/functions.rst [hunk 5]
```diff
@@ -282,3 +282,25 @@ function follows:
   is used when performing a ``GROUP BY`` aggregation, and an implementation
   will be automatically generated for you, if you don't specify a
   ``AccumulatorStateFactory``
+
+Deprecated Function
+-------------------
+
+The ``@Deprecated`` annotation can be used to mark a function as deprecated
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/develop/functions.rst]: deprecated. It should

File: presto-docs/src/main/sphinx/develop/functions.rst [hunk 6]
```diff
@@ -282,3 +282,25 @@ function follows:
...
+and should no longer be used. Presto will generates a warning whenever a SQL
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/develop/functions.rst]: Presto generates

Comment (mosabua) [reply presto-docs/src/main/sphinx/develop/functions.rst]: and SQL statement

File: presto-docs/src/main/sphinx/develop/functions.rst [hunk 7]
```diff
@@ -282,3 +282,25 @@ function follows:
...
+uses a deprecated function. When a function is deprecated, it should also be
```
Comment (mosabua) [inline presto-docs/src/main/sphinx/develop/functions.rst]: When a functions is deprecated, the ``@Description`` needs to be replaced with a note about the deprecation and the replacement function:

**Comment by mosabua** [comment]
Thank you for that contribution. This is great. 

In terms of rolling this out to users it might make sense to do the following:

- ship with feature flag and default to OFF for a release or two, and document this in release notes and normal documentation
- ship with feature flag and default to ON afterwards for better overall usability
- potentially remove feature flag and leave always on after depending on community input

**Comment by mosabua** [comment]
Also .. I review the doc based on request from @martint

File: presto-docs/src/main/sphinx/develop/functions.rst (hunk 6 above)
Comment (oneonestar) [reply presto-docs/src/main/sphinx/develop/functions.rst]: > and SQL statement

You mean `a SQL statement`?
```diff
-The ``@Deprecated`` annotation can be used to mark a function as deprecated
-and should no longer be used. Presto will generates a warning whenever a SQL
+The ``@Deprecated`` annotation can be used to mark a function as deprecated.
+It should no longer be used. Presto generates a warning whenever a SQL statement
```

File: presto-docs/src/main/sphinx/develop/functions.rst [hunk 8]
```diff
@@ -282,3 +282,25 @@ function follows:
   is used when performing a ``GROUP BY`` aggregation, and an implementation
   will be automatically generated for you, if you don't specify a
   ``AccumulatorStateFactory``
+
+Deprecated Function
+-------------------
+
+The ``@Deprecated`` annotation can be used to mark a function as deprecated.
+It should no longer be used. Presto generates a warning whenever a SQL statement
```
Comment (martint) [inline presto-docs/src/main/sphinx/develop/functions.rst]: This is a little ambiguous. It's not clear whether it refers to the function or the annotation. Maybe rephrase as:

> The ``@Deprecated`` annotation can be used to mark a function as deprecated and to indicate that it should no longer be used.

@mosabua, do you have any suggestions?

File: presto-main/src/main/java/io/prestosql/SystemSessionProperties.java [hunk 9]
```diff
@@ -523,6 +524,11 @@ public SystemSessionProperties(
                         "Experimental: Use late materialization (including WorkProcessor pipelines)",
                         featuresConfig.isLateMaterializationEnabled(),
                         false),
+                booleanProperty(
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/SystemSessionProperties.java]: I don't think this is necessary. This change does not affect behavior or break compatibility, so there's no need to make it configurable. It also won't be "active" unless a function is already deprecated.

File: presto-main/src/main/java/io/prestosql/operator/scalar/JsonFunctions.java [hunk 10]
```diff
@@ -360,17 +361,21 @@ public static Boolean jsonArrayContains(@SqlType(StandardTypes.JSON) Slice json,
...
     {
         return jsonArrayGet(json, index);
     }
 
+    @Deprecated
     @SqlNullable
     @ScalarFunction
+    @Description("(DEPRECATED) If the extracted element is a string, it will be converted into an invalid JSON")
     @SqlType(StandardTypes.JSON)
     public static Slice jsonArrayGet(@SqlType(StandardTypes.JSON) Slice json, @SqlType(StandardTypes.BIGINT) long index)
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/operator/scalar/JsonFunctions.java]: This should be a separate change. Also, we should discuss whether we want to deprecate those functions or fix their behavior. So, let's start with an issue that describes how the behavior is broken and go from there.

File: presto-main/src/main/java/io/prestosql/sql/analyzer/FeaturesConfig.java [hunk 11]
```diff
@@ -965,6 +966,18 @@ public FeaturesConfig setLateMaterializationEnabled(boolean lateMaterializationE
...
+        return deprecatedFunctionWarningEnabled;
+    }
+
+    @Config("analyzer.deprecated-function-warning-enabled")
+    public FeaturesConfig setDeprecatedFunctionWarningEnabled(boolean deprecatedFunctionWarningEnabled)
+    {
+        this.deprecatedFunctionWarningEnabled = deprecatedFunctionWarningEnabled;
+        return this;
+    }
+
```
Comment (martint) [inline presto-main/src/main/java/io/prestosql/sql/analyzer/FeaturesConfig.java]: Same here... I don't think this is necessary. This change does not affect behavior or break compatibility, so there's no need to make it configurable.

File: presto-tests/src/test/java/io/prestosql/execution/TestUDFPluginDeprecatedUDFPlugin.java [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+import com.google.common.collect.ImmutableSet;
+import io.prestosql.spi.Plugin;
+import io.prestosql.spi.function.Description;
+import io.prestosql.spi.function.ScalarFunction;
+import io.prestosql.spi.function.SqlType;
+import io.prestosql.spi.type.StandardTypes;
+
+import java.util.Set;
+
+public class TestUDFPluginDeprecatedUDFPlugin
```
Comment (martint) [inline presto-tests/src/test/java/io/prestosql/execution/TestUDFPluginDeprecatedUDFPlugin.java]: This class name has too many redundant words. How about calling it `DeprecatedFunctionsPlugin`?

File: presto-docs/src/main/sphinx/develop/functions.rst (hunk 8 above)
Comment (mosabua) [reply presto-docs/src/main/sphinx/develop/functions.rst]: haha.. that was my suggestion .. I thought it is clear as one sentence. 
Maybe we use

The ``@Deprecated`` annotation has to be used on any function that should no longer be used. The annotation causes Presto to generate a warning whenever...

File: presto-testing/src/main/java/io/prestosql/testing/DistributedQueryRunner.java [hunk 13]
```diff
@@ -278,6 +279,12 @@ public void installPlugin(Plugin plugin)
         log.info("Installed plugin %s in %s", plugin.getClass().getSimpleName(), nanosSince(start).convertToMostSuccinctTimeUnit());
     }
 
+    @Override
+    public void addFunctions(List<? extends SqlFunction> functions)
+    {
+        servers.forEach(s -> s.getMetadata().addFunctions(functions));
```
Comment (martint) [inline presto-testing/src/main/java/io/prestosql/testing/DistributedQueryRunner.java]: Rename `s` to `server`

File: presto-main/src/main/java/io/prestosql/operator/scalar/JsonFunctions.java (hunk 10 above)
Comment (oneonestar) [reply presto-main/src/main/java/io/prestosql/operator/scalar/JsonFunctions.java]: Ref: #2573

**Comment by martint** [comment]
Merged, thanks!
